src/main/java/bio/terra/workspace/service/job/JobService.java

```diff
@@ -118,6 +120,16 @@ protected String submit(
       Class<? extends Flight> flightClass, FlightMap parameterMap, String jobId) {
     try {
       stairway.submit(jobId, flightClass, parameterMap);
+    } catch (DatabaseOperationException dbEx) {
+      String sqlState = (((SQLException) dbEx.getCause()).getSQLState());
+      // sqlState 23505 indicates a unique key violation, which in this case indicates a duplicate
+      // flightId. See https://www.postgresql.org/docs/10/errcodes-appendix.html for postgres
+      // error codes.
+      if (sqlState.equals("23505")) {
+        throw new DuplicateFlightIdException("Flight ID already exists in Stairway: " + jobId);
+      } else {
+        throw new InternalStairwayException(dbEx);
```

Comment:
Agree that this is pretty hacky atm. I think the problem with checking for the ID before trying to submit is that this is inside our concurrency control logic - we can't be sure another request with the same ID isn't being processed at the same time. 
I think my ideal solution to this would be a checked exception or a distinct return value from Stairway on duplicate flightID. I'll check with Dan about those ideas later today.

Reply:
I don't know that we need to overly worry about the race between the id exists & resubmitting. If we do have concurrent requests that try to submit the same id, one of them will fail thanks to Stairway's guarantee at submission time, so no harm no foul. But any retry will work correctly. So it'd be a little nice for clients to never have the funny 503 error, we do get eventual correctness. I think the more common retry case is that the response to the first request gets lost. The second request won't be concurrent, so will work directly.
+1 checking with Dan.

Reply:
Followup on this, I'm moving the duplicate flightID check to be a checked exception from Stairway in https://github.com/DataBiosphere/stairway/pull/42. I'll update WM here when that gets merged and published.